pkg/autodiscovery/providers/cloudfoundry.go

```diff
@@ -0,0 +1,206 @@
+	cfp := CloudFoundryConfigProvider{
+		lastCollected: time.Now(),
+	}
+	var err error
+
+	cfp.bbsCache, err = cloudfoundry.GetGlobalBBSCache()
+	if err != nil {
+		return nil, err
+	}
+	return cfp, nil
+}
+
+// String returns a string representation of the CloudFoundryConfigProvider
+func (cf CloudFoundryConfigProvider) String() string {
+	return names.CloudFoundryBBS
+}
+
+// IsUpToDate returns true if the last collection time was later than last BBS Cache refresh time
+func (cf CloudFoundryConfigProvider) IsUpToDate() (bool, error) {
+	return cf.lastCollected.After(cf.bbsCache.LastUpdated()), nil
+}
+
+// Collect collects AD config templates from all relevant BBS API information
+func (cf CloudFoundryConfigProvider) Collect() ([]integration.Config, error) {
+	log.Debugf("Collecting configs via the CloudFoundryProvider")
+	cf.lastCollected = time.Now()
+	allActualLRPs, desiredLRPs := cf.bbsCache.GetAllLRPs()
+	allConfigs := []integration.Config{}
+	for _, desiredLRP := range desiredLRPs {
+		actualLRPs, ok := allActualLRPs[desiredLRP.AppGUID]
+		if !ok {
+			actualLRPs = []cloudfoundry.ActualLRP{}
+		}
+		newConfigs := cf.getConfigsForApp(desiredLRP, actualLRPs)
+		log.Debugf("Successfully got %d configs for app %s", len(newConfigs), desiredLRP.AppGUID)
+		allConfigs = append(allConfigs, newConfigs...)
+	}
+	return allConfigs, nil
+}
+
+func (cf CloudFoundryConfigProvider) getConfigsForApp(desiredLRP cloudfoundry.DesiredLRP, actualLRPs []cloudfoundry.ActualLRP) []integration.Config {
+	allConfigs := []integration.Config{}
+
+	for adName, adVal := range desiredLRP.EnvAD {
+		// initially, let's assume a non-container service; we'll change to container service in
+		// `expandPerContainerChecks` if necessary
+		id := cloudfoundry.NewADNonContainerIdentifier(desiredLRP, adName)
+		// we need to convert adVal to map[string]string to pass it to extractTemplatesFromMap
+		convertedADVal := map[string]string{}
+		for k, v := range adVal {
+			convertedADVal[k] = string(v)
+		}
+		parsedConfigs, errs := extractTemplatesFromMap(id.String(), convertedADVal, "")
+		for _, err := range errs {
+			log.Errorf("Cannot parse endpoint template for service %s of app %s: %s, skipping",
+				adName, desiredLRP.AppGUID, err)
+		}
+
+		vcVal, vcOk := desiredLRP.EnvVcapServices[adName]
+		variables, varsOk := adVal["variables"]
+		success := false
+		switch {
+		case vcOk:
+			// if service is found in VCAP_SERVICES (non-container service), we will run a single check per App
+			err := cf.renderExtractedConfigs(parsedConfigs, variables, vcVal)
+			if err != nil {
+				log.Errorf("Failed to render config for service %s of app %s: %s", adName, desiredLRP.AppGUID, err)
+			} else {
+				success = true
+			}
+			cf.assignNodeNameToNonContainerChecks(parsedConfigs, desiredLRP, actualLRPs)
+		case varsOk:
+			log.Errorf("Service %s for app %s has variables configured, but is not present in VCAP_SERVICES", adName, desiredLRP.AppGUID)
```

Comment:
Should we log existing vcap services for debug purpose?

Reply:
Yeah, good point. Will do.

Reply:
Aren't there credentials in there ?

Reply:
It's only the services names, ignore me :)